### PR TITLE
Validate local development scenarios in CI

### LINKED ISSUE #56237: Validate local development workflows in CI
While a given change may pass all required CI checks, it could still silently break one or more aspects of local development. To prevent that from happening, we should consider adding a CI pipeline that simulates common local development workflows. This pipeline would serve two purposes:

1. Ensure that local development workflows remain intact
2. Perform basic sanity checks about how the repo is getting built

For example, we could use this pipeline to verify that the locally built ASP.NET Core shared framework and packages are getting picked up by sample/test projects within this repository. By keeping these checks straightforward, any failures can be treated as real issues. We would never disable or quarantine these checks.

This proposal is motivated from a recent change in build infrastructure that disrupted the local development workflow for testing web project templates (the automated tests were also broken in CI, but this went unnoticed because the tests were already being skipped). Some areas of the repo aren't touched by developers frequently, so a new CI pipeline could help prevent the development experience of these areas from gradually degrading over time.

### PR DESCRIPTION
**Pull request by MackinnonBuck**
Adds a check to ensure that local development scenarios are not broken:
* Build and pack (Debug)
* Build and pack (Release)
* Run project template tests
  * These depend on the locally built shared framework being available
  * Currently _all_ project template tests are run, but we could speed up this check a bit by limiting this to a smaller subset of tests (e.g., Blazor tests only)

This check can be further extended with additional scenarios as we see fit.

Fixes #56237

### CODE REVIEW

File: .azure/pipelines/ci.yml [hunk 1]
```diff
@@ -701,6 +701,38 @@ extends:
               publishOnError: true
               includeForks: true
 
+        # Local development validation
```
Comment (wtgodbe) [inline .azure/pipelines/ci.yml:704]: If you want this in PR CI, it needs to be in `ci-public.yml`

File: .azure/pipelines/ci.yml [hunk 2]
```diff
@@ -701,6 +701,40 @@ extends:
...
+        - template: .azure/pipelines/jobs/default-build.yml@self
```
Comment (wtgodbe) [inline .azure/pipelines/ci.yml:705]: I'd recommend adding a condition so this only runs on PRs, similar to Code Check: https://github.com/dotnet/aspnetcore/blob/63c492e22b06d4903cb4e7aee037295c71e1ec37/.azure/pipelines/ci.yml#L158-L159

**Comment by MackinnonBuck** [comment]
Looks like we're required to specify a minor and patch version when installing the JDK. This is hardcoded right now, as it is [in helix](https://github.com/dotnet/aspnetcore/blob/main/eng/targets/Helix.targets#L3-L4). See also: https://github.com/dotnet/aspnetcore/issues/57476

**Review by wtgodbe — APPROVED** [review]
LGTM, but can you run an internal build against this branch to confirm the ci.yml changes didn't bust anything? Just push the branch to https://dev.azure.com/dnceng/internal/_git/dotnet-aspnetcore & queue a build of https://dev.azure.com/dnceng/internal/_build?definitionId=21 against that branch

**Comment by MackinnonBuck** [comment]
The internal build passed; I'm going to merge this now.

**Pull request merged**